opm/common/utility/ThreadedMap.hpp

```diff
@@ -0,0 +1,127 @@
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+#ifndef THREADED_MAP_HPP
+#define THREADED_MAP_HPP
+
+#include <algorithm>
+#include <cstddef>
+#include <list>
+#include <numeric>
+#include <unordered_map>
+#include <vector>
+
+#if _OPENMP
+#include <omp.h>
+#endif
+
+namespace Opm {
+
+//! \brief Utility class for handling map building in parallel.
+//! details Intended to be used for building maps in OpenMP loops.
+//!         In the parallel section the values are assembled into a
+//!         list per thread. After the threaded section the lists are
+//!         assembled to the map, either through RAII in the dtor,
+//!         or through a manual call to finalize()
+
+template<class Map>
+class ThreadedMap
+{
+    //! \brief Predicate for maps
+    template<class T>
+    struct is_unordered
+    {
+        constexpr static bool value = false;
+    };
+
+    template<class Key, class T, class Compare, class Allocator>
+    struct is_unordered<std::unordered_map<Key,T,Compare,Allocator>>
+    {
+        constexpr static bool value = true;
+    };
+
+public:
+    using Key = typename Map::key_type;
+    using Value = typename Map::mapped_type;
+
+    //! \brief Constructor for a given map and number of threads.
+    ThreadedMap(Map& map, int num_threads)
+        : map_(map), num_threads_(num_threads)
+    {
+#if _OPENMP
+        if (num_threads_ > 1) {
+            thread_values_.resize(num_threads_);
+        }
+#endif
+    }
+
+    //! \brief The destructor (optionally) assembles the final map.
+    ~ThreadedMap()
+    {
+        finalize();
+    }
+
+    //! \brief Insert a value into the map.
+    void emplace(Key key, Value value)
+    {
+#if _OPENMP
+        if (num_threads_ > 1) {
+            thread_values_[omp_get_thread_num()].
+                emplace_back(std::move(key), std::move(value));
+        } else
+#endif
+        {
+            map_.emplace(std::move(key), std::move(value));
+        }
+    }
+
+    //! \brief Assembles the final map.
+    void finalize()
+    {
+#if _OPENMP
+        if (num_threads_ > 1 && !thread_values_.empty()) {
+            if constexpr (is_unordered<Map>::value) {
+                const std::size_t size =
+                    std::accumulate(thread_values_.begin(),
+                                    thread_values_.end(), std::size_t(0),
+                                [](std::size_t s, const auto& v) { return s + v.size(); });
+                    map_.reserve(size);
+            }
+            for (int i = 0; i < num_threads_; ++i) {
+                for (const auto& it : thread_values_[i]) {
+                    map_.emplace(std::move(it.first), std::move(it.second));
+                }
```

Comment:
If multiple threads input the same value, the last thread wins. This might be what you want, but only in very special cases.

Reply:
Thanks for making me think about this, even though your statement is wrong. As-is first thread wins, I want last thread wins.